graphql_compiler/compiler/workarounds/orientdb_query_execution.py

```diff
@@ -0,0 +1,341 @@
+
+    - Ineligible locations are ones that cannot be the starting point of query execution.
+    - Preferred locations are ones that are eligible to be the starting point, and also have
+      an associated "where:" clause that references no non-local fields -- only local fields,
+      literals, and variables.
+    - Eligible locations are all locations that do not fall into either of these two categories.
+
+    Args:
+        match_query: MatchQuery object describing the query being analyzed for optimization
+
+    Returns:
+        tuple (preferred, eligible, ineligible) where each element is a set of Location objects.
+        The three sets are disjoint.
+    """
+    preferred_locations = set()
+    eligible_locations = set()
+    ineligible_locations = set()
+
+    # Any query must have at least one traversal with at least one step.
+    # The first step in this traversal must be a QueryRoot.
+    first_match_step = match_query.match_traversals[0][0]
+    if not isinstance(first_match_step.root_block, QueryRoot):
+        raise AssertionError(u'First step of first traversal unexpectedly was not QueryRoot: '
+                             u'{} {}'.format(first_match_step, match_query))
+
+    # The first step in the first traversal cannot possibly be inside an optional, recursion,
+    # or fold. Its location is always an eligible start location for a query.
+    # We need to determine whether it is merely eligible, or actually a preferred location.
+    if first_match_step.where_block is not None and _is_local_filter(first_match_step.where_block):
+        preferred_locations.add(first_match_step.as_block.location)
+    else:
+        eligible_locations.add(first_match_step.as_block.location)
+
+    # This loop will repeat the analysis of the first step of the first traversal.
+    # QueryRoots other than the first are required to always be at a location whose status
+    # (preferred / eligible / ineligible) is already known. Since we already processed
+    # the first QueryRoot above, the rest of the loop can assume all QueryRoots are like that.
+    for current_traversal in match_query.match_traversals:
+        for match_step in current_traversal:
+            current_step_location = match_step.as_block.location
+
+            already_encountered_location = any((
+                current_step_location in preferred_locations,
+                current_step_location in eligible_locations,
+                current_step_location in ineligible_locations,
+            ))
+
+            if isinstance(match_step.root_block, QueryRoot):
+                if not already_encountered_location:
+                    raise AssertionError(u'Unexpectedly encountered a location in QueryRoot whose '
+                                         u'status has not been determined: {} {} {}'
+                                         .format(current_step_location, match_step, match_query))
+
+                at_eligible_location = (
```

Comment:
Similar issue as above. Here, it seems like `at_eligible_location` is referring to both eligible and preferred locations.

Reply:
Maybe change this one to `at_eligible_or_preferred_location`?